src/main/java/org/elasticsearch/script/ScriptContext.java

```diff
@@ -19,20 +19,62 @@
 
 package org.elasticsearch.script;
 
-import java.util.Locale;
-
 /**
- * Operation/api that uses a script as part of its execution.
- * Note that the suggest api is considered part of search for simplicity, as well as the percolate api.
+ * Context of an operation that uses scripts as part of its execution.
  */
-public enum ScriptContext {
-    MAPPING,
-    UPDATE,
-    SEARCH,
-    AGGS;
-
-    @Override
-    public String toString() {
-        return name().toLowerCase(Locale.ROOT);
+public interface ScriptContext {
+
+    /**
+     * @return the name of the operation
+     */
+    String key();
+
+    /**
+     * Standard operations that make use of scripts as part of their execution.
+     * Note that the suggest api is considered part of search for simplicity, as well as the percolate api.
+     */
+    enum Standard implements ScriptContext {
+
+        AGGS("aggs"), MAPPING("mapping"), SEARCH("search"), UPDATE("update");
+
+        private final String key;
+
+        Standard(String key) {
+            this.key = key;
+        }
+
+        @Override
+        public String key() {
+            return key;
+        }
+    }
+
+    /**
+     * Generic custom operation exposed via plugin
+     *
+     * @deprecated create a new {@link org.elasticsearch.script.ScriptContext.Plugin} instance instead
+     */
+    @Deprecated
+    Plugin GENERIC_PLUGIN = new Plugin("plugin");
+
+    /**
+     * Custom operation exposed via plugin, which makes use of scripts as part of its execution
+     */
+    class Plugin implements ScriptContext {
+
+        private final String key;
+
+        /**
+         * Creates a new custom scripts based operation exposed via plugin
+         * @param key the name of the operation. can be used to enable/disable scripts via fine-grained settings
+         */
+        public Plugin(String key) {
```

Comment:
how will you be able to detect duplicates? hmm... maybe I'm missing something here... the way I saw it, each plugin will effectively define a static its contexts (probably static constants of `SearchContext.Plugin`) and then pass it to executable/compile on demand